chainer/functions/pooling/roi_align_2d.py

```diff
@@ -0,0 +1,529 @@
+# Modified work:
+# -----------------------------------------------------------------------------
+# Copyright (c) 2018 Preferred Infrastructure, Inc.
+# Copyright (c) 2018 Preferred Networks, Inc.
+# -----------------------------------------------------------------------------
+
+# Original work:
+# -----------------------------------------------------------------------------
+# Copyright (c) 2015 by Contributors
+# \file roi_pooling.cu
+# \brief roi pooling operator
+# \author Ross Girshick, Kye-Hyeon Kim, Jian Guo
+# \changed to roi_align by Elaine Bao
+# \file roi_align.cu
+# \roi align operator described in Mask RCNN
+# -----------------------------------------------------------------------------
+
+import collections
+
+import numpy
+import six
+
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _pair(x):
+    if isinstance(x, collections.Iterable):
+        return x
+    return x, x
+
+
+def _get_bilinear_interp_params(y, x, height, width):
+    if y < -1 or y > height or x < -1 or x > width:
+        # out of range, so it is empty
+        return (None,) * 8
+
+    if y <= 0:
+        y = 0
+    if x <= 0:
+        x = 0
+
+    y_low = int(y)
+    x_low = int(x)
+
+    if y_low >= height - 1:
+        y_high = y_low = height - 1
+        y = float(y_low)
+    else:
+        y_high = y_low + 1
+
+    if x_low >= width - 1:
+        x_high = x_low = width - 1
+        x = float(x_low)
+    else:
+        x_high = x_low + 1
+
+    ly = y - y_low
+    lx = x - x_low
+    hy = 1. - ly
+    hx = 1. - lx
+
+    w1 = hy * hx
+    w2 = hy * lx
+    w3 = ly * hx
+    w4 = ly * lx
+
+    return y_low, x_low, y_high, x_high, w1, w2, w3, w4
+
+
+_GET_BILINEAR_INTERP_KERNEL = '''
```

Comment:
I think it's better to define a device function instead of inserting strings into the kernel code. You can use the device function from the kernel by giving the device function code to `preamble` argument of `ElementwiseKernel`. Please see the following code and check what I'm saying

How to define the device function:
https://gist.github.com/mitmul/bf9db024adfa5a39508d74bc00745d28#file-roi_align_2d-py-L72-L121

How to use the function inside of the elementwise kernel code:
https://gist.github.com/mitmul/bf9db024adfa5a39508d74bc00745d28#file-roi_align_2d-py-L304-L314

What you need to do to insert the device function code on top of the `.cu` file generated by `ElementwiseKernel`:
https://gist.github.com/mitmul/bf9db024adfa5a39508d74bc00745d28#file-roi_align_2d-py-L335